arch/xtensa/core/syscall_helper.c

```diff
@@ -27,3 +28,22 @@ uintptr_t xtensa_syscall_helper(uintptr_t arg1, uintptr_t arg2,
 
 	return a2;
 }
+#endif /* CONFIG_XTENSA_SYSCALL_USE_HELPER */
+
+#if XCHAL_HAVE_THREADPTR == 0
+#include <xtensa/config/core-isa.h>
+#include <xtensa/config/core.h>
+
+bool xtensa_is_user_context(void)
+{
+	uint32_t ret;
+
+	__asm__ volatile(".global xtensa_is_user_context_epc\n"
+			 "        xtensa_is_user_context_epc:\n"
+			 "                syscall\n"
+			 "                mov %0, a2\n"
+			 : "=r"(ret) : : "a2");
+
```

Comment:
Starting to think that we should always use `syscall` for this.  It would simplify a lot the code having only one way to do this. With this change we have three different ways:
1) Using `THREADPTR` directly
2) Using TLS
3) With syscall

Reply:
The obvious downside is speed. So I guess we can have syscall and TLS, and removing THREADPTR entirely. Though we can do it in future commits if decided to do so.